### PR TITLE
add ProcessBlockHeader v0.6

### PR DESCRIPTION
**Pull request by shayzluf**
[Part of] #2307 

---

# Add ProcessBlockHeader

includes:
ProcessBlockHeader

**Write why you are making the changes in this pull request**
to comply with 0.6 spec

**Write a summary of the changes you are making**
changes in:
- beacon-chain/core/blocks/block_operations.go

**Link anything that would be helpful or relevant to the reviewers**
https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#block-header

### CODE REVIEW

File: proto/beacon/p2p/v1/types.proto [hunk 1]
```diff
@@ -167,16 +167,17 @@ message Crosslink {
 }
 
 message BeaconBlock {
+  // Deprecated
+  // TODO(2307): Remove deprecated fields
+  bytes randao_reveal = 1003;
+  Eth1Data eth1_data = 1004;
+  // 0.6 fields
```
Comment (terencechain) [inline proto/beacon/p2p/v1/types.proto]: remove 0.6 fields

File: proto/beacon/p2p/v1/types.proto [hunk 2]
```diff
@@ -167,16 +167,17 @@ message Crosslink {
 }
 
 message BeaconBlock {
+  // Deprecated
```
Comment (terencechain) [inline proto/beacon/p2p/v1/types.proto]: why did you change the ordering, please revert

File: beacon-chain/core/blocks/block_operations.go [hunk 3]
```diff
@@ -68,6 +69,84 @@ func ProcessEth1DataInBlock(beaconState *pb.BeaconState, block *pb.BeaconBlock)
...
+	//verify proposer is not slashed
+	idx,err:=helpers.BeaconProposerIndex(beaconState)
+	if err!=nil{
+		return nil,err
+	}
+	proposer:= beaconState.ValidatorRegistry[idx]
+	if proposer.Slashed{
+		return nil,fmt.Errorf("proposer id: %d was slashed",idx)
+	}
+	// TODO(#2307) reapply after bls.Verify is finished
```
Comment (terencechain) [inline beacon-chain/core/blocks/block_operations.go]: no commented code please, please take this out

File: beacon-chain/core/blocks/block_operations.go [hunk 4]
```diff
@@ -68,6 +69,84 @@ func ProcessEth1DataInBlock(beaconState *pb.BeaconState, block *pb.BeaconBlock)
...
+	bHash,err:=ssz.TreeHash(bBytes)
+	if err!=nil {
+		return nil,err
+	}
+	beaconState.LatestBlockHeader = &pb.BeaconBlockHeader{
+		Slot: block.Slot,
+		PreviousBlockRoot: block.ParentBlockRoot,
+		BlockBodyRoot: bHash[:],
+	}
+	//verify proposer is not slashed
```
Comment (terencechain) [inline beacon-chain/core/blocks/block_operations.go]: new line here and format
```suggestion

	// Verify proposer is not slashed
```

File: beacon-chain/core/blocks/block_operations.go [hunk 5]
```diff
@@ -4,6 +4,7 @@
 package blocks
 
 import (
+	"github.com/prysmaticlabs/prysm/shared/ssz"
```
Comment (terencechain) [inline beacon-chain/core/blocks/block_operations.go]: format

File: proto/beacon/p2p/v1/types.proto (hunk 2 above)
Comment (shayzluf) [reply proto/beacon/p2p/v1/types.proto]: i am dependent on field order for self_signed function

File: beacon-chain/core/blocks/block_operations.go (hunk 3 above)
Comment (shayzluf) [reply beacon-chain/core/blocks/block_operations.go]: when bls.Verify will be ready i just need to uncomment it

Comment (terencechain) [reply beacon-chain/core/blocks/block_operations.go]: It is usually not encouraged, reason being when `bls.Verify` finishes, it might be different & obsolete commented code block. I think this is ok given we are on `0.6` branch and things are moving fairly fast

File: proto/beacon/p2p/v1/types.proto (hunk 1 above)
Comment (shayzluf) [reply proto/beacon/p2p/v1/types.proto]: replaced with // End of deprecated fields

File: beacon-chain/core/blocks/block_operations.go (hunk 4 above)
Comment (shayzluf) [reply beacon-chain/core/blocks/block_operations.go]: thanks

File: beacon-chain/core/blocks/block_operations.go (hunk 5 above)
Comment (shayzluf) [reply beacon-chain/core/blocks/block_operations.go]: thanks

File: beacon-chain/core/blocks/block_operations.go (hunk 3 above)
Comment (shayzluf) [reply beacon-chain/core/blocks/block_operations.go]: great, thanks

File: proto/beacon/p2p/v1/types.proto (hunk 2 above)
Comment (shayzluf) [reply proto/beacon/p2p/v1/types.proto]: reverted

File: proto/beacon/p2p/v1/types.proto (hunk 1 above)
Comment (shayzluf) [reply proto/beacon/p2p/v1/types.proto]: reverted

File: beacon-chain/core/blocks/block_operations.go [hunk 6]
```diff
@@ -68,6 +69,85 @@ func ProcessEth1DataInBlock(beaconState *pb.BeaconState, block *pb.BeaconBlock)
...
+//     # Verify proposer signature
+//     assert bls_verify(proposer.pubkey, signing_root(block), block.signature, get_domain(state, DOMAIN_BEACON_PROPOSER))
+func ProcessBlockHeader(
+	beaconState *pb.BeaconState,
+	block *pb.BeaconBlock,
+) (*pb.BeaconState, error) {
+	if beaconState.Slot != block.Slot {
+		return nil, fmt.Errorf("state slot: %d is different then block slot: %d", beaconState.Slot, block.Slot)
+	}
+	lbhsr, err := ssz.SignedRoot(beaconState.LatestBlockHeader)
```
Comment (nisdas) [inline beacon-chain/core/blocks/block_operations.go]: please use better variable naming here

File: beacon-chain/core/blocks/block_operations_test.go [hunk 7]
```diff
@@ -1825,3 +1825,296 @@ func TestProcessValidatorExits_AppliesCorrectStatus(t *testing.T) {
 			helpers.DelayedActivationExitEpoch(state.Slot/params.BeaconConfig().SlotsPerEpoch), newRegistry[0].ExitEpoch)
 	}
 }
+
+func TestProcessBlockHeader_OK(t *testing.T) {
+	t.Skip()
+	// TODO(#2307) unskip after bls.Verify is finished
```
Comment (nisdas) [inline beacon-chain/core/blocks/block_operations_test.go]: why is this skipped ?, bls.Verify is commented out

File: shared/ssz/ssz_utils_cache.go [hunk 8]
```diff
@@ -110,3 +110,20 @@ func structFields(typ reflect.Type) (fields []field, err error) {
 	}
 	return fields, nil
 }
+
+func structFieldMap(typ reflect.Type) (fieldMap map[string]field, err error) {
```
Comment (nisdas) [inline shared/ssz/ssz_utils_cache.go]: can you unit test this too ?

File: beacon-chain/core/helpers/randao.go [hunk 9]
```diff
@@ -80,5 +80,6 @@ func RandaoMix(state *pb.BeaconState, wantedEpoch uint64) ([]byte, error) {
 		return nil, fmt.Errorf("input randaoMix epoch %d out of bounds: %d <= epoch < %d",
 			wantedEpoch, earliestEpoch, currentEpoch)
 	}
-	return state.LatestRandaoMixes[wantedEpoch%params.BeaconConfig().LatestRandaoMixesLength], nil
+	idx := wantedEpoch % params.BeaconConfig().LatestRandaoMixesLength
```
Comment (nisdas) [inline beacon-chain/core/helpers/randao.go]: check if the index is valid here, basically `if idx >- len(state.LatestRandaoMixes) `

File: beacon-chain/core/blocks/block_operations_test.go (hunk 7 above)
Comment (shayzluf) [reply beacon-chain/core/blocks/block_operations_test.go]: removed

File: beacon-chain/core/helpers/randao.go (hunk 9 above)
Comment (shayzluf) [reply beacon-chain/core/helpers/randao.go]: added test and if

Comment (nisdas) [reply beacon-chain/core/helpers/randao.go]: i dont think you pushed your latest changes up

File: shared/ssz/ssz_utils_cache.go (hunk 8 above)
Comment (shayzluf) [reply shared/ssz/ssz_utils_cache.go]: added tests

File: beacon-chain/core/helpers/randao.go [hunk 10]
```diff
@@ -80,5 +80,9 @@ func RandaoMix(state *pb.BeaconState, wantedEpoch uint64) ([]byte, error) {
...
+	if idx >= uint64(len(state.LatestRandaoMixes)) {
+		return nil, fmt.Errorf("requested id in latest randao mix is out of bound requested id: %v, array length: %v", idx, len(state.LatestRandaoMixes))
```
Comment (nisdas) [inline beacon-chain/core/helpers/randao.go]: Can you put it as index or idx instead of id

File: proto/beacon/p2p/v1/types.proto [hunk 11]
```diff
@@ -172,7 +172,7 @@ message BeaconBlock {
   bytes state_root = 3;
   BeaconBlockBody body = 4;
   bytes signature = 5; // bytes96
-
+
```
Comment (nisdas) [inline proto/beacon/p2p/v1/types.proto]: Revert the space

File: shared/ssz/ssz_utils_cache_test.go [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+	blim byte
+}
+type fieldTest struct {
+	val      interface{}
+	index    int
+	sigFound bool
+	error    string
+}
+
+func TestStructFieldMap_OK(t *testing.T) {
```
Comment (nisdas) [inline shared/ssz/ssz_utils_cache_test.go]: Nice tests :)

File: beacon-chain/core/blocks/block_operations.go (hunk 6 above)
Comment (shayzluf) [reply beacon-chain/core/blocks/block_operations.go]: fixed

File: beacon-chain/core/helpers/randao.go (hunk 10 above)
Comment (shayzluf) [reply beacon-chain/core/helpers/randao.go]: changed

File: proto/beacon/p2p/v1/types.proto (hunk 11 above)
Comment (shayzluf) [reply proto/beacon/p2p/v1/types.proto]: done

File: shared/ssz/ssz_utils_cache_test.go (hunk 12 above)
Comment (shayzluf) [reply shared/ssz/ssz_utils_cache_test.go]: thanks

File: beacon-chain/rpc/beacon_server.go [hunk 13]
```diff
@@ -248,6 +248,7 @@ func (bs *BeaconServer) BlockTree(ctx context.Context, _ *ptypes.Empty) (*pb.Blo
 	}, nil
 }
 
+//nolint
```
Comment (nisdas) [inline beacon-chain/rpc/beacon_server.go]: why ?

File: beacon-chain/core/epoch/epoch_processing.go [hunk 14]
```diff
@@ -18,16 +18,8 @@ import (
 	"github.com/prysmaticlabs/prysm/shared/params"
 	"github.com/prysmaticlabs/prysm/shared/sliceutil"
 	"github.com/prysmaticlabs/prysm/shared/ssz"
-	"github.com/sirupsen/logrus"
 )
 
-var log = logrus.WithField("prefix", "core/state")
```
Comment (nisdas) [inline beacon-chain/core/epoch/epoch_processing.go]: why was this removed

File: beacon-chain/core/epoch/epoch_processing.go [hunk 15]
```diff
@@ -18,16 +18,8 @@ import (
...
-
-type queueElement struct {
```
Comment (nisdas) [inline beacon-chain/core/epoch/epoch_processing.go]: also this

File: beacon-chain/core/epoch/epoch_processing.go (hunk 14 above)
Comment (shayzluf) [reply beacon-chain/core/epoch/epoch_processing.go]: the linter cried that it is not used

File: beacon-chain/core/epoch/epoch_processing.go (hunk 15 above)
Comment (shayzluf) [reply beacon-chain/core/epoch/epoch_processing.go]: also lint error

Comment (shayzluf) [reply beacon-chain/core/epoch/epoch_processing.go]: not used

File: beacon-chain/core/epoch/epoch_processing.go (hunk 14 above)
Comment (nisdas) [reply beacon-chain/core/epoch/epoch_processing.go]: ok , sounds good then

File: beacon-chain/rpc/beacon_server.go (hunk 13 above)
Comment (shayzluf) [reply beacon-chain/rpc/beacon_server.go]: function not used in ci

File: beacon-chain/core/blocks/block_operations.go [hunk 16]
```diff
@@ -59,6 +59,85 @@ func ProcessEth1DataInBlock(beaconState *pb.BeaconState, block *pb.BeaconBlock)
...
+) (*pb.BeaconState, error) {
+	if beaconState.Slot != block.Slot {
+		return nil, fmt.Errorf("state slot: %d is different then block slot: %d", beaconState.Slot, block.Slot)
+	}
+	latestBlockHeaderSR, err := ssz.SignedRoot(beaconState.LatestBlockHeader)
+	if err != nil {
+		return nil, err
+	}
+	if !bytes.Equal(block.ParentBlockRoot, latestBlockHeaderSR[:]) {
+		return nil, fmt.Errorf("state parentblockroot: %d is different then latest block header signed root: %d",
```
Comment (terencechain) [inline beacon-chain/core/blocks/block_operations.go]:
```suggestion
		return nil, fmt.Errorf("state parent block root: %d is different then latest block header signed root: %d",
```

Comment (terencechain) [reply beacon-chain/core/blocks/block_operations.go]: why `%d`?

File: beacon-chain/core/blocks/block_operations.go [hunk 17]
```diff
@@ -59,6 +59,85 @@ func ProcessEth1DataInBlock(beaconState *pb.BeaconState, block *pb.BeaconBlock)
...
+	}
+
+	// Verify proposer is not slashed
+	idx, err := helpers.BeaconProposerIndex(beaconState)
+	if err != nil {
+		return nil, err
+	}
+	proposer := beaconState.ValidatorRegistry[idx]
+	if proposer.Slashed {
+		return nil, fmt.Errorf("proposer id: %d was slashed", idx)
```
Comment (terencechain) [inline beacon-chain/core/blocks/block_operations.go]:
```suggestion
		return nil, fmt.Errorf("proposer index %d was slashed", idx)
```

File: beacon-chain/core/blocks/block_operations_test.go [hunk 18]
```diff
@@ -10,14 +10,15 @@ import (
 	"testing"
 	"time"
 
+	"github.com/prysmaticlabs/prysm/shared/hashutil"
+
```
Comment (terencechain) [inline beacon-chain/core/blocks/block_operations_test.go]: format

File: beacon-chain/rpc/beacon_server.go (hunk 13 above)
Comment (terencechain) [reply beacon-chain/rpc/beacon_server.go]: why not just delete it?

File: shared/ssz/ssz_utils_cache.go [hunk 19]
```diff
@@ -110,3 +110,23 @@ func structFields(typ reflect.Type) (fields []field, err error) {
...
```
Comment (terencechain) [inline shared/ssz/ssz_utils_cache.go]: what does this have to do with `ProcessBlockHeader`?

Comment (rauljordan) [reply shared/ssz/ssz_utils_cache.go]: It's part of ssz.SigningRoot which is used in there

File: beacon-chain/rpc/beacon_server.go (hunk 13 above)
Comment (shayzluf) [reply beacon-chain/rpc/beacon_server.go]: removed

File: beacon-chain/core/blocks/block_operations.go (hunk 16 above)
Comment (shayzluf) [reply beacon-chain/core/blocks/block_operations.go]: changed to %#x

File: beacon-chain/core/blocks/block_operations.go (hunk 17 above)
Comment (shayzluf) [reply beacon-chain/core/blocks/block_operations.go]: thanks

File: beacon-chain/core/blocks/block_operations_test.go (hunk 18 above)
Comment (shayzluf) [reply beacon-chain/core/blocks/block_operations_test.go]: thanks
